File: java02t/src/step11/ex01/Test03.java
package step11.ex01;

public class Test03 {
  
  // Singleton 클래스 만들기
  //1) 외부에서 생성할 수 없도록 생성자를 private으로 접근을 제한한다.
  //2) 인스턴스를 리턴해 줄 스태틱 메서드를 만든다.
  //   - 인스턴스 주소를 저장할 클래스 변수를 만든다.
  //   - 한 번 인스턴스를 생성하면 다음 번 부터는 기존의 인스턴스를 리턴한다.
  
  public static void main(String[] args) {
    // Singleton 적용 후 
    // => getInstance()가 리턴하는 것은 같은 인스턴스다.
    Calculator c1 = Calculator.getInstance();
    Calculator c2 = Calculator.getInstance();
    
    c1.plus(10);
    c1.plus(20);
    c1.minus(3);
    
    c2.plus(30);
    c2.minus(40);
    
    System.out.printf("%d, %d\n", c1.result, c2.result);
  }
}

File: node01t/src/step01/test03.js
// 함수 선언과 사용
var a = 100;
var b = 20;

// 변수에 함수 객체를 할당하는 것은 호이스팅 대상이 아니다.
// 따라서 아래의 함수는 사용하기 전에 정의해야 한다.
var minus = function(v1, v2) {
	return v1 - v2;
};

console.log("a + b = ", plus(a, b));
console.log("a - b = ", minus(a, b));

// 호이스팅(hoisting) 때문에 아래의 함수 선언은 자동으로 코드의 맨 위로 올라간다.
// 따라서 아래 함수를 사용하는 코드가 함수 정의 위에 있어도 된다.
function plus(v1, v2) {
	return v1 + v2;
}
